### PR TITLE
feat(config-endpoint): add initial implementation

### PR DESCRIPTION
**Pull request by etolbakov**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

add `/config` API endpoint to retrieve the current configuration(Options) that were used for the app to start

Please explain IN DETAIL what the changes are in this PR and why they are needed:

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link
https://github.com/GreptimeTeam/greptimedb/issues/1793

### CODE REVIEW

**Comment by etolbakov** [comment]
Hello @killme2008 @sunng87 @MichaelScofield @waynexia 
Could I please have your feedback on the approach that I took for this task?

I understand the main challenge is to pass the `Options` enum to the http.
My idea is to use a file system for info exchange, particularly, write `Options` on start into `/tmp/greptimedb/conf/opts.json` as an example (hardcoded at the moment).
When it comes to `/config` API invocation, it will boil down to reading a file from that directory.
1) Does this approach seem fine for you and should I proceed? Please let me know if there's a better way of doing that?
2) In the original describption the desired output format was `toml`. Do we consider using `json`?

**Comment by waynexia** [comment]
>In the original describption the desired output format was toml. Do we consider using json?

I prefer toml because it's the format we (the binary) accept 🧐

**Comment by MichaelScofield** [comment]
@etolbakov I think `/config` endpoint could just return plain text in http, does not need to use an intermediate file to exchange data. As to the output format, I strongly suggest `toml`. So that user could copy the output from `/config` endpoint and paste them in another GreptimeDB, very convenient, very ergonomic.

**Comment by etolbakov** [comment]
Thank you very much for the feedback 👍!

Michael, sorry for confusion. As one remarkable programmer told - "Talk is cheap, show me the code", so I put a bit more details to support my initial idea. Could you please take a look when you have time?

Here's how it works at the moment. _The created file:_
```sh
❯ ls -lah /tmp/greptimedb/conf
drwxr-xr-x  3 etolbakov  wheel    96B  6 Jul 14:45 .
drwxr-xr-x  8 etolbakov  wheel   256B  6 Jul 15:04 ..
-rw-r--r--  1 etolbakov  wheel   1.8K  6 Jul 15:04 opts.toml    
```
_The app start log:_
```
❯ /target/release/greptime standalone start
2023-07-06T14:04:26.845441Z  INFO greptime: short_version: 0.3.2, full_version: greptimedb-feat/http-config-endpoint-c60c695
2023-07-06T14:04:26.845543Z  INFO greptime: command line arguments
2023-07-06T14:04:26.845547Z  INFO greptime: argument: ./target/release/greptime
2023-07-06T14:04:26.845559Z  INFO greptime: argument: standalone
2023-07-06T14:04:26.845561Z  INFO greptime: argument: start
2023-07-06T14:04:26.846407Z  INFO greptime: Configuration options were successfully written at '/tmp/greptimedb/conf/opts.toml'
2023-07-06T14:04:26.846501Z  INFO cmd::standalone: Standalone start command: StartCommand {
    http_addr: None,
.... <redacted>
```
 _Endpoint invocation example:_
```
❯ curl  http://127.0.0.1:4000/config
type = "Standalone"
type = "MixOptions"

[fe_opts]
mode = "standalone"
heartbeat_interval_millis = 5000
retry_interval_millis = 5000

[fe_opts.http_options]
addr = "127.0.0.1:4000"
timeout = "30s"
body_limit = "64MiB"
.... <redacted>
```
Happy to proceed with adding tests and putting it under admin api if there are no major reservations with this approach.

File: src/cmd/src/bin/greptime.rs [hunk 1]
```diff
@@ -235,3 +235,29 @@ async fn main() -> Result<()> {
...
     Ok(())
 }
+
+/// Default path to store configuration
+const DEFAULT_OPTIONS_PATH: &str = "/tmp/greptimedb/conf/opts.toml";
+
+fn write_opts_into_tmp_dir(opts: &Options) {
+    let opts_toml = toml::to_string(&opts).unwrap();
+
+    if let Some(parent_dir) = std::path::Path::new(DEFAULT_OPTIONS_PATH).parent() {
```
Comment (etolbakov) [inline src/cmd/src/bin/greptime.rs:245]: path could be taken from the env variable and `DEFAULT_OPTIONS_PATH` could be used if a variable is not specified

**Comment by MichaelScofield** [comment]
@etolbakov I see, so you are generating a static file from config at the start of running GreptimeDB, and reading it back to user in http text when `/config` endpoint is called. I'm ok with that. However, is it simpler to just store the config struct in http service, then upon be called `/config`, **dynamically** render it to http text and return? The frequency of calling endpoint `/config` won't be too much, and the convertion (config struct -> toml file format) should be cheap, so I think the cost is fine.

**Comment by etolbakov** [comment]
@MichaelScofield, thank you very much for the feedback!
Your comment made me to review possible options again, I think I came up with slightly better approach.
Could you please take a look when you have time?

File: src/servers/src/http.rs [hunk 2]
```diff
@@ -518,6 +530,14 @@ impl HttpServer {
             routing::get(handler::health).post(handler::health),
         );
 
+        let config_router = self
+            .route_config(GreptimeOptionsConfigState {
+                greptime_config_options: self.greptime_config_options.clone().unwrap(),
+            })
+            .finish_api(&mut api);
+
+        router = router.nest(&format!("/{HTTP_API_VERSION}"), config_router);
```
Comment (etolbakov) [inline src/servers/src/http.rs:539]: > Remember to put these administration APIs under a different path from data APIs. We don't expose these APIs in some deployment.

Do I understand it correctly that `"/{HTTP_API_VERSION}/admin"`is required here?

Comment (MichaelScofield) [reply src/servers/src/http.rs:539]: If it's admin API, I don't think it should be placed under "version" path. Other admin APIs like "/health", "/status" or "/metrics" don't.

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -5,7 +5,7 @@ on:
   schedule:
     # At 00:00 on Monday.
     - cron: '0 0 * * 1'
-  # Mannually trigger only builds binaries.
+  # Manually trigger only builds binaries.
```
Comment (etolbakov) [inline .github/workflows/release.yml:8]: fix a typo I've noticed back in a day, not related to the scope of this PR

**Comment by etolbakov** [comment]
@MichaelScofield @waynexia 
Could you please take a look when you have a spare minute?

File: tests-integration/tests/http.rs [hunk 4]
```diff
@@ -495,6 +496,78 @@ pub async fn test_health_api(store_type: StorageType) {
     assert_eq!(body, HealthResponse {});
 }
 
+pub async fn test_config_api(store_type: StorageType) {
```
Comment (waynexia) [inline tests-integration/tests/http.rs:499]: 👍

**Review by waynexia — APPROVED** [review]
Looking good, thanks

**Review by MichaelScofield — APPROVED** [review]

**Pull request merged**